File: 14.解决循环引用/14.解决循环引用/ViewController.m
//
//  ViewController.m
//  14.解决循环引用
//
//  Created by 浅爱 on 16/3/7.
//  Copyright © 2016年 y. All rights reserved.
//

#import "ViewController.h"
#import "NetworkTools.h"


@interface ViewController ()

@property (nonatomic, strong) NetworkTools *tools;

@end

@implementation ViewController

- (void)viewDidLoad {
    
    [super viewDidLoad];

    
    self.tools = [[NetworkTools alloc] init];
    
    // 使用__weak 解决block的循环引用
    // 对象销毁后，weakSELF指向的地址为nil
    // ios5.0 提供
//     __weak typeof(self) weakSelf = self;
    
    // 使用__unsafe_unretained 解决block的循环引用
    // iOS4.0 提供
    // 对象销毁后，weakSelf指向地址不变，地址对应对象已经销毁了，崩溃了 报野指针
    __unsafe_unretained typeof(self) weakSelf = self;
    
    [self.tools loadData:^(NSString *result) {
        
        NSLog(@"%@", result);
        
        NSLog(@"%@", weakSelf.view);
        
    }];

}


-(void)dealloc {

    NSLog(@"ViewController dealloc");

}






- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

@end

File: 04.可选项类型/04.可选项类型/ViewController.swift
//
//  ViewController.swift
//  04.可选项类型
//
//  Created by 浅爱 on 16/3/5.
//  Copyright © 2016年 y. All rights reserved.
//

import UIKit

class ViewController: UIViewController {

    override func viewDidLoad() {
        super.viewDidLoad()

        demo1()
    }
    
    
    
    
    func demo1() {
    
        let a: Int? = 10
        
        let b = 20
        
        if a != nil {
        
            let result = a! + b
            
            print(result)
        
        }
        
        let result = (a ?? 0) + b
        
        print(result)
        
        
        let c: Int? = nil
        
        print(c)
        
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }


}

